### ISSUE TITLE
dom() for download not responding - working folder changed halfway

### ISSUE THREAD

**Issue by yangliu8912** [tebelorg/TagUI-Python#60]
Hi Ken,

I am again using TagUI for a downloading task. This time the download does not happen from a url, I need to call DownloadFile(fileID,groupID).

So I used this in TagUI:

t.dom("DownloadFile(10, 3894)")
print("hello")
t.dom("DownloadFile(20008, 1311)")

The first file DownloadFile(10, 3894) is downloaded successfully, but it hangs there. The second call is not invoked at all, not even the "hello" is printed. What could be wrong?

Best regards,
Yang Liu

**Comment by kensoh** [comment]
Hi Yang Liu, I'm assuming for your website there is no other way to download - for example using t.click() or using t.download() to download directly from the URL?

Sending JavaScript code to the web browser DOM to download something is not common, I assume above scenario you are invoking a function in the webpage to do that.

Can you check the following - 

1. Try doing this in Chrome dev tools window to see if running this there downloads ok and control returns back to the dev tools window.

2. Run with t.debug(True) and attach the tagui_python.log log file here to see if there are some clues.

3. Lastly, check the folder where TagUI is installed and attach the tagui/src/tagui_chrome.log here to see if there are clues. For Windows, location of installation is %APPDATA%\tagui and for macOS / Linux it is in ~/.tagui folder.

**Comment by yangliu8912** [comment]
Hi Ken,

Thank you for giving me the direction to dig into the issue. 

My python script was actually together with my csv file in another directory, say "myDirectory". I actually set the working directory to myDirectory to first read the csv into a dataframe, so that I can loop through a long list to invoke the DownloadFile(fileID,groupID) function.

As I learnt from my previous issue #56 
TagUI process is actually invoked from C:\Users\liuyang1
I just experimented by moving my python script and csv from myDirectory into C:\Users\liuyang1
such that I do not have to change directory in order to take in the csv file.

The t.dom("DownloadFile(fileID,groupID)") worked fine after I did this.

I tested again for the scenario where I had to change working directory, it just won't respond...

I still don't quite understand why changing the working directory matters, but it magically solved my problem.

Thank you.

Best regards,
Yang Liu

**Comment by kensoh** [comment]
That's great!

TagUI will use the location of the script or the location where the Python process was launched as the default download direction. This setting is made when you access an URL using t.url(). 

Try changing the current working directory before you initialise TagUI and it should work. For example the following works. The root cause is TagUI for Python looks for an internal file at the current working directory (the location when it was launched). If the folder changes halfway, TagUI for Python will keep looking for the internal I/O file at the new current working directory. 

But that file is not there, because the TagUI process is still writing to the original folder on launch. It is possible to modify code to handle the edge case of changing working directory halfway, but because the code section will impact all the main use cases, I'll not make this change for now, but raise an issue with this question and answer so that people changing directory halfway and hangs will know why.

```py
import os
import tagui as t

# folder changed before invoking TagUI
os.chdir('test')

t.init()
t.url('https://github.com/tebelorg/TagUI-Python')
t.click('Clone or download')
t.click('Download ZIP')
t.wait()
t.close()

# zip file downloaded to the changed folder 'test'
# instead of the original Python script folder
```

If doing it above way still does not work, possible reasons could be the user for your Python or Jupyter notebook process does not have the write access to that directory.

**Comment by kensoh** [comment]
Don't think avoiding a challenge is good spirit, will figure out if there's a way to do this without risking existing users' scripts compatibility and behaviour.

**Comment by kensoh** [comment]
Closing issue here for now, to work on the active #62 to explore a way to make this work.

**Comment by kensoh** [comment]
Fyi Yang Liu, a change is implemented to support changing directory halfway.

Implemented in v1.13, and now available with `pip install tagui --upgrade`
